envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/EnvoyListenersFactory.kt

```diff
@@ -105,21 +115,52 @@ class EnvoyListenersFactory(
                                         .setAddress(listenersConfig.ingressHost)
                         )
                 )
-                .addFilterChains(createIngressFilterChain(group, listenersConfig, globalSnapshot))
-                .build()
+
+        listOfNotNull(securedIngressChain, insecureIngressChain).forEach {
+            listener.addFilterChains(it.build())
+        }
+
+        return listener.build()
     }
 
     private fun createIngressFilterChain(
         group: Group,
         listenersConfig: ListenersConfig,
-        globalSnapshot: GlobalSnapshot
-    ): FilterChain {
-
-        val filterChain = FilterChain.newBuilder()
+        globalSnapshot: GlobalSnapshot,
+        transportProtocol: String = "raw_buffer"
+    ): FilterChain.Builder {
+        return FilterChain.newBuilder()
+                .setFilterChainMatch(FilterChainMatch.newBuilder().setTransportProtocol(transportProtocol))
                 .addFilters(createIngressFilter(group, listenersConfig, globalSnapshot))
+    }
 
-        return filterChain
+    private fun createSecuredIngressFilterChain(
+        group: Group,
+        listenersConfig: ListenersConfig,
+        globalSnapshot: GlobalSnapshot
+    ): FilterChain.Builder {
+        val filterChain = createIngressFilterChain(group, listenersConfig, globalSnapshot, "tls")
+        val tlsAuthenticationProperties = snapshotProperties.incomingPermissions.tlsAuthentication
+
+        val downstreamTlsContext = DownstreamTlsContext.newBuilder()
+                .setRequireClientCertificate(BoolValue.of(true)) // if false, will not validate cert
+                .setCommonTlsContext(CommonTlsContext.newBuilder()
+                        .addTlsCertificateSdsSecretConfigs(SdsSecretConfig.newBuilder()
```

Comment:
We need to:
* set TLS version to 1.2
* set cipher suites to:
```
ECDHE-ECDSA-AES128-GCM-SHA256
ECDHE-RSA-AES128-GCM-SHA256
```

It's been a while since we performed a benchmark. In the meantime, default TLS version changed in Envoy from 1.2 to 1.3. Maybe they fixed performance of TLS 1.3 and we should run benchmark again...

https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#extensions-transport-sockets-tls-v3-tlsparameters